Fix nested lists

diunko opened the pull request:
That fix looks more like a hack to me.
blockEnd doesn't work like expected for listItem -- it terminates all containing lists, so the fact that embedded lists work in homogeneous case (ul within ul or ol within ol) is more like a coincidence ;) 
It either needs some refactoring or is misunderstood by me.

diunko committed:
add failing test

diunko committed:
add a fix

arikon commented:
@afelix Could you make code review?

@diunko Could you provide that refactoring? ;)

diunko commented:
It's on the way to cost an arm and a leg to a tiny file.

afelix commented:
Дим, а добавь ещё тест, пожалуйста, на 4-уровневый список вида "список4 в списке3 в списке2 в списке1".

diunko committed:
test case for fourfold embedded lists

diunko committed:
Refactor embedded lists. A real fix this time.

Hopefully cleaner implementation allows to track newlines
and spaces consumption a bit easier.

arikon commented on src/shmakowiki.ometajs:
```diff
@@ -68,40 +68,52 @@ ometa ShmakoWiki {
-
-    list :t = ~~(apply(t):s) listItem(t, s)+:c -> [t + 'st', c],
+    listItem :t :n =
+      ' '*:s ?(n==s.length) bullet1(t)
```
Code style: `n == s.length`. And other places also.

diunko committed:
coding style bump

diunko committed:
other places coding style bump also

diunko commented:
Anything else strange in style?

diunko committed:
bump indent

diunko committed:
Bump style. Add more space. Code can breath.

diunko committed:
typo

arikon commented:
@veged @indutny Hey guys, please have a look

veged commented on tests/tests.js:
```diff
@@ -236,6 +236,66 @@ var tests = [
+            '      * uli2223\n'+
+            '    3. oli223\n'+
+            '  * uli23\n'+
+            '3. li3\n'),
```
maybe `.join('\n')`?

veged commented on src/shmakowiki.ometajs:
```diff
@@ -68,40 +68,52 @@ ometa ShmakoWiki {
 
-    anyLi = uli | oli,
+    bullet = uliBullet | oliBullet,
+    bullet1 :t = apply(t+'liBullet'),
```
I prefer space after `+` quantifier

veged commented:
seems very nice!

can merge after few little polishing things in comments

diunko committed:
requested style fixes
